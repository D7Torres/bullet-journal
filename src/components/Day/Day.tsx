import { addItem, getItemsByDate } from '@api/items'
import type { Item } from '@api/items'
import { toFriendlyWeekDate } from '@utils/dates'
import { useCallback, useEffect, useState } from 'react'
import debounce from 'lodash.debounce'

export const Day = ({ date }: { date: Date }) => {
  const [items, setItems] = useState<Item[]>([])
  const [isEditing, setIsEditing] = useState(false)
  const [isSaving, setIsSaving] = useState(false)

  useEffect(() => {
    const items = getItemsByDate(date)
    console.log('>>>', { items })
    setItems(items)
  }, [date])

  const handleItemOnChange = (item: Item, newText: string) => {
    setIsSaving(true)
    saveItemDebounced({ ...item, text: newText })
  }

  const saveItemDebounced = useCallback((item: Item) => {
    debounce(() => {
      addItem(item)
      setIsSaving(false)
    }, 500)()
  }, [])

  return (
    <>
      <h2>{toFriendlyWeekDate(date)}</h2>
      {!isEditing ? (
        <>
          <button onClick={() => setIsEditing(true)}>Write</button>
          <ul>
            {items.map((item) => (
              <li key={item.id}>{item.text}</li>
            ))}
          </ul>
        </>
      ) : (
        <>
          <button disabled={isSaving} onClick={() => setIsEditing(false)}>
            I'm done
          </button>
          <ul>
            {items.map((item) => (
              <li key={item.id}>
                <input
                  onChange={(e) => handleItemOnChange(item, e.target.value)}
                  type="text"
                  defaultValue={item.text}
                />
              </li>
            ))}
          </ul>
        </>
      )}
    </>
  )
}
