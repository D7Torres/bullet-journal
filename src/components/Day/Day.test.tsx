import { render } from '@testing-library/react'
import { vi } from 'vitest'
import { Day } from './Day'
import { describe } from 'node:test'
import type { Item } from '@api/items'
import '@testing-library/jest-dom'

const date = new Date('2021-01-01')
const otherDate = new Date('2021-01-02')
const mockItems: Record<string, Item[]> = {
  '2021-01-01': [
    { id: 1, date, text: 'Item 1', isSoftDeleted: false },
    { id: 2, date, text: 'Item 2', isSoftDeleted: false },
  ],
  '2021-01-02': [
    {
      id: 3,
      date: otherDate,
      text: 'Item belonging to a different date',
      isSoftDeleted: false,
    },
  ],
}

describe('Day component', () => {
  beforeAll(() => {
    vi.mock('@api/items', () => ({
      getItemsByDate: (date: Date) => {
        const dateString = date.toISOString().split('T')[0]

        return mockItems[dateString]
      },
    }))
  })

  test('renders', () => {
    const { asFragment } = render(<Day date={date} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('items on that day are visible', () => {
    const { getByText } = render(<Day date={date} />)

    expect(getByText('Item 1')).toBeInTheDocument()
    expect(getByText('Item 2')).toBeInTheDocument()
  })

  test('items on other days are not visible', () => {
    const { queryByText } = render(<Day date={date} />)

    expect(
      queryByText('Item belonging to a different date'),
    ).not.toBeInTheDocument()
  })
})
