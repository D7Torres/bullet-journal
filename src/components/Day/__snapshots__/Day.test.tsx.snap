// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Day component > renders 1`] = `
<DocumentFragment>
  <h2>
    Friday, 1
  </h2>
  <button>
    Write
  </button>
  <ul>
    <li>
      Item 1
    </li>
    <li>
      Item 2
    </li>
  </ul>
</DocumentFragment>
`;
